### PR TITLE
WIP: avoid zip and unzip to improve the speed of get_indexer

### LINKED ISSUE #23735: PERF: regression in reindex. Pandas 0.23.4 is 60x slower than 0.22.0 with a MultiIndex with datetime64
Re-indexing a series with a two-level MultiIndex where the first level is datetime64 and the second level is int is 40x slower than in 0.22.0. Output first then repro code below. The issue persists if you change the first level to int instead of datetime, but the perf difference is less (0.40 seconds vs 0.03 seconds).

```python

"""
pandas version: 0.23.4
reindex took 1.9770500659942627 seconds

pandas version: 0.22.0
reindex took 0.0306899547577 seconds
"""


import pandas as pd
import time
import numpy as np


if __name__ == '__main__':
    n_days = 300
    dr = pd.date_range(end="20181118", periods=n_days)
    mi = pd.MultiIndex.from_product([dr, range(1440)])

    v = np.random.randn(len(mi))
    mask = np.random.rand(len(v)) < .3
    v[mask] = np.nan
    s = pd.Series(v, index=mi)
    s = s.sort_index()

    s2 = s.dropna()

    start = time.time()

    s2.reindex(index=s.index)

    end = time.time()
    print("pandas version: %s" % pd.__version__)
    print("reindex took %s seconds" % (end - start))

```

### PR DESCRIPTION
**Pull request by qiuwei**
- [x] closes #23735 #34531 
- [ ] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] improve the speed for get_index by avoiding some redundant zipping and unzipping

### CODE REVIEW

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -715,6 +715,24 @@ def _values(self) -> np.ndarray:
...
+    def _values_for_indexer(self):
+        values = []
+
+        for i in range(self.nlevels):
+            vals = self._get_level_values(i)
+            if is_categorical_dtype(vals.dtype):
+                vals = cast("CategoricalIndex", vals)
+                vals = vals._data._internal_get_values()
+            if isinstance(vals.dtype, ExtensionDtype):
+                vals = vals.astype(object)
```
Comment (jbrockmendel) [inline pandas/core/indexes/multi.py:728]: for e.g. dt64tz wouldnt we want to just use dt64?

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -2666,6 +2684,15 @@ def _get_partial_string_timestamp_match_key(self, key):
 
         return key
 
+    def _get_engine_target(self) -> np.ndarray:
```
Comment (jbrockmendel) [inline pandas/core/indexes/multi.py:2687]: annotation is inaccurate?

**Comment by datapythonista** [comment]
@qiuwei is this something you want to continue working on?

**Comment by qiuwei** [comment]
> @qiuwei is this something you want to continue working on?

@datapythonista I realize there are a lot of implementations depending on the current `get_indexer` interface.
Changing this requires too much work for me at the moment.  

Close for now.

**Pull request closed without merging**